arangod/Aql/EnumerateListExecutor.cpp

```diff
@@ -70,84 +72,149 @@ RegisterId EnumerateListExecutorInfos::getOutputRegister() const noexcept {
+  THROW_ARANGO_EXCEPTION(TRI_ERROR_NOT_IMPLEMENTED);
+}
 
-    if (!_currentRow.isInitialized()) {
-      TRI_ASSERT(_rowState == ExecutionState::DONE);
-      return {_rowState, NoStats{}};
-    }
+void EnumerateListExecutor::initializeNewRow(AqlItemBlockInputRange& inputRange) {
+  if (_currentRow) {
+    std::ignore = inputRange.nextDataRow();
+  }
+  std::tie(_currentRowState, _currentRow) = inputRange.peekDataRow();
+  if (!_currentRow) {
+    return;
+  }
 
-    AqlValue const& inputList = _currentRow.getValue(_infos.getInputRegister());
+  // fetch new row, put it in local state
+  AqlValue const& inputList = _currentRow.getValue(_infos.getInputRegister());
 
-    if (_inputArrayPosition == 0) {
-      // store the length into a local variable
-      // so we don't need to calculate length every time
-      if (inputList.isDocvec()) {
-        _inputArrayLength = inputList.docvecSize();
-      } else {
-        if (!inputList.isArray()) {
-          throwArrayExpectedException(inputList);
-        }
-        _inputArrayLength = inputList.length();
-      }
+  // store the length into a local variable
+  // so we don't need to calculate length every time
+  if (inputList.isDocvec()) {
+    _inputArrayLength = inputList.docvecSize();
+  } else {
+    if (!inputList.isArray()) {
+      throwArrayExpectedException(inputList);
     }
+    _inputArrayLength = inputList.length();
+  }
+
+  _inputArrayPosition = 0;
+}
+
+void EnumerateListExecutor::processArrayElement(OutputAqlItemRow& output) {
+  bool mustDestroy;
+  AqlValue const& inputList = _currentRow.getValue(_infos.getInputRegister());
+  AqlValue innerValue = getAqlValue(inputList, _inputArrayPosition, mustDestroy);
+  AqlValueGuard guard(innerValue, mustDestroy);
+
+  TRI_IF_FAILURE("EnumerateListBlock::getSome") {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+  }
 
-    if (_inputArrayLength == 0) {
+  output.moveValueInto(_infos.getOutputRegister(), _currentRow, guard);
+  output.advanceRow();
+
+  // set position to +1 for next iteration after new fetchRow
+  _inputArrayPosition++;
+}
+
+size_t EnumerateListExecutor::skipArrayElement(size_t toSkip) {
+  size_t skipped = 0;
+
+  if (toSkip <= _inputArrayLength - _inputArrayPosition) {
+    // if we're skipping less or exact the amount of elements we can skip with toSkip
+    _inputArrayPosition += toSkip;
+    skipped = toSkip;
+  } else if (toSkip > _inputArrayLength - _inputArrayPosition) {
+    // we can only skip the max amount of values we've in our array
+    skipped = _inputArrayLength - _inputArrayPosition;
+    _inputArrayPosition = _inputArrayLength;
+  }
+  return skipped;
+}
+
+std::tuple<ExecutorState, NoStats, AqlCall> EnumerateListExecutor::produceRows(
+    AqlItemBlockInputRange& inputRange, OutputAqlItemRow& output) {
+  AqlCall upstreamCall{};
+  upstreamCall.fullCount = output.getClientCall().fullCount;
+
+  if (!inputRange.hasDataRow()) {
+    return {inputRange.upstreamState(), NoStats{}, upstreamCall};
+  }
```

Comment:
I am always confused with that... what if the output was full and thus the current array has elements left. In that case I think you would receive the same input row containing the data again as you only consume the row when you wrote the array completely?